torch/distributed/fsdp/_wrap_utils.py

```diff
@@ -119,3 +127,138 @@ def _warn_on_overridden_mixed_precision(
+                msg += (
+                    f"The following parameters have requires_grad=True:\n{nonfrozen_param_fqns}\n"
+                    f"The following parameters have requires_grad=False:\n{frozen_param_fqns}"
+                )
+                if use_orig_params:
+                    warnings.warn(msg)
+                else:
+                    raise ValueError(msg)
+
+
+def _get_topo_sorted_named_modules(root_module: nn.Module) -> List[nn.Module]:
+    """
+    We use a stack-based DFS instead of ``root_module.named_modules()`` to get
+    the topological order since reversing the DFS order gives the modules in
+    registration order at each level in the module tree (as opposed to in
+    reverse), which allows us to error/warn on the first registered module that
+    violates the condition.
+
+    For example, consider the following module structure:
+        M(
+          S1(),
+          S2(
+            SS1(),
+            SS2(),
+          ),
+          S3(),
+        )
+    The reverse DFS order is [S1, SS1, SS2, S2, S3, M], while the reverse
+    ``named_modules()`` order is [S3, SS2, SS1, S2, S1, M].
+    """
+    visited_modules = {root_module}
+    stack = [("", root_module)]
+    topo_sorted_named_modules: List[Tuple[str, nn.Module]] = []
+    while stack:
+        module_name, module = stack.pop()
+        topo_sorted_named_modules.append((module_name, module))
+        for child_module_name, child_module in module.named_children():
+            if child_module is None:  # only for overrides of `named_children()`
+                continue
+            if child_module not in visited_modules:
+                visited_modules.add(child_module)
+                if module_name != "":
+                    child_module_name = module_name + "." + child_module_name
+                stack.append((child_module_name, child_module))
+    return topo_sorted_named_modules
+
+
+def _get_managed_param_to_fqn(
+    module_to_wrap: nn.Module,
+    ignored_params: Set[nn.Parameter],
+    visited_modules: Set[nn.Module],
+    root_prefix: str,
+) -> Dict[nn.Parameter, str]:
+    """
+    This returns a dict that maps managed parameter to its FQN for the given
+    ``module_to_wrap``. The dict's keys are exactly the parameters that would
+    be managed by the module, where this is achieved by calling this function
+    on the modules to wrap in reverse topological order, destructively updating
+    ``visited_modules``, and not traversing into those modules. The FQNs are
+    prefixed from the root (via ``root_prefix``) to be more informative.
+
+    NOTE: This function is meant to be called pre-wrapping and iteratively in
+    reverse topological order to cover the full module tree. This differs from
+    the ``_get_param_to_fqn()`` function meant to be called post-wrapping and
+    on the full module tree in one shot. Given those differences, we do not try
+    to unify the two.
```

Comment:
This function really smacks of "we want to know what the module tree would look like after we do wrapping, but we need to run this before we actually start wrapping, so we have to simulate what the effect would be." That's annoying, because if you change how the wrap works, you also have to adjust this function too (speaking of which, if this statement is true, it would be helpful to have some invariant, like for example `_get_param_to_fqn(post_wrapping_module)` should report the same as `_get_managed_param_to_fqn(pre_wrapping_module)`--I'm not sure if this holds, but it would be helpful to know if it is.)